server/src/main/scala/com/cloudera/livy/server/ApiVersioningSupport.scala

```diff
@@ -0,0 +1,98 @@
+package com.cloudera.livy.server
+
+import javax.servlet.http.HttpServletRequest
+
+import org.scalatra.{NotAcceptable, ScalatraBase}
+
+/**
+  * Livy's servlets can mix-in this trait to get API version support.
+  *
+  * Example: {{{
+  * import ApiVersions._
+  * class FooServlet
+  *   ...
+  *   with ApiVersioningSupport
+  *   ...
+  * {
+  *   get("/test") {
+  *     ...
+  *   }
+  *   get("/test", getApiVersion() <= v0_2) {
+  *     ...
+  *   }
+  *   get("/test", getApiVersion() <= v0_1) {
+  *     ...
+  *   }
+  * }
+  * }}}
+  */
+trait ApiVersioningSupport extends AbstractApiVersioningSupport {
+  this: ScalatraBase =>
+  // Link the abstract trait to Livy's version enum.
+  override val apiVersions = ApiVersions
+  override type ApiVersionType = ApiVersions.Value
+}
+
+trait AbstractApiVersioningSupport {
+  this: ScalatraBase =>
+  protected val apiVersions: Enumeration
+  protected type ApiVersionType
+
+  /**
+    * Before proceeding with routing, validate the specified API version in the request.
+    * If validation passes, cache the parsed API version as a per-request attribute.
+    */
+  before() {
+    try {
+      request(AbstractApiVersioningSupport.ApiVersionKey) = parseForApiVersion(request)
+    } catch {
+      case e: Exception =>
+        halt(NotAcceptable(e.getMessage))
+    }
+  }
+
+  /**
+    * @return The specified API version in the request.
+    */
+  def getApiVersion(): ApiVersionType = {
+    request(AbstractApiVersioningSupport.ApiVersionKey).asInstanceOf[ApiVersionType]
+  }
+
+  private def extractVersionFromAcceptHeader(acceptHeader: String): Option[String] = {
+    // Get every character after "application/vnd.livy.v" until hitting a + sign.
+    val acceptHeaderRegex = """application/vnd\.livy\.v([^\+]*).*""".r
+
+    acceptHeader match {
+      case acceptHeaderRegex(v) => Option(v)
+      case _ => None
+    }
+  }
+
+  private def parseForApiVersion(request: HttpServletRequest): ApiVersionType = {
```

Comment:
Sorry I'm new to Scala and haven't grasped how powerful pattern matching is.

I fixed the pattern matching. But I want to keep parseApiVersion() purely a parsing function so I left error reporting unchanged and kept `request(AbstractApiVersioningSupport.ApiVersionKey) = version` out of parseApiVersion().